qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -116,77 +140,202 @@ def __init__(self, hls_config=None):
+            # we use the list of methods provided by the user
+            methods = self.hls_config.methods[op.name]
+        elif (
+            self.hls_config.use_default_on_unspecified
+            and "default" in hls_plugin_manager.method_names(op.name)
+        ):
+            # the operation's name does not appear in the user-specified config,
+            # we use the "default" method when instructed to do so and the "default"
+            # method is available
+            methods = ["default"]
+        else:
+            methods = []
+
+        for method in methods:
+            # There are two ways to specify a synthesis method. The more explicit
+            # way is to specify it as a tuple consisting of a synthesis algorithm and a
+            # list of additional arguments, e.g.,
+            #   ("kms", {"all_mats": 1, "max_paths": 100, "orig_circuit": 0}), or
+            #   ("pmh", {}).
+            # When the list of additional arguments is empty, one can also specify
+            # just the synthesis algorithm, e.g.,
+            #   "pmh".
+            if isinstance(method, tuple):
+                plugin_specifier, plugin_args = method
+            else:
+                plugin_specifier = method
+                plugin_args = {}
+
+            # There are two ways to specify a synthesis algorithm being run,
+            # either by name, e.g. "kms" (which then should be specified in entry_points),
+            # or directly as a class inherited from HighLevelSynthesisPlugin (which then
+            # does not need to be specified in entry_points).
+            if isinstance(plugin_specifier, str):
+                if plugin_specifier not in hls_plugin_manager.method_names(op.name):
+                    raise TranspilerError(
+                        "Specified method: %s not found in available plugins for %s"
+                        % (plugin_specifier, op.name)
+                    )
+                plugin_method = hls_plugin_manager.method(op.name, plugin_specifier)
+            else:
+                plugin_method = plugin_specifier
+
+            # ToDo: similarly to UnitarySynthesis, we should pass additional parameters
+            #       e.g. coupling_map to the synthesis algorithm.
+            # print(f"{plugin_method = }, {op = }, {plugin_args = }")
+            decomposition = plugin_method.run(op, **plugin_args)
+
+            # The synthesis methods that are not suited for the given higher-level-object
+            # will return None, in which case the next method in the list will be used.
+            if decomposition is not None:
+                return decomposition
+
+        return None
+
+    def _synthesize_annotated_op(self, op: Operation) -> Union[Operation, None]:
+        """
+        Recursively synthesizes annotated operations.
+        Returns either the synthesized operation or None (which occurs when the operation
+        is not an annotated operation).
+        """
```

Comment:
I agree with that, and I was thinking to later add an option that skips synthesizing annotated operations. I very much like the suggestion to put annotated operations synthesis under a plugin, then we would easily have multiple methods to handle these, including fully synthesizing (as done right now) or skipping. Since the plugin mechanism is name-based, does it makes sense that the `name` of an annotated operation is always going to be `"annotated"` (which is how it's right now), or do we want it to be more descriptive (potentially showing the modifiers and the base operation, somewhat akin to names of controlled gates)?